hatch/app.py

```diff
@@ -50,18 +52,30 @@ def validate(request: Request, auth_token: str = Security(api_key_header)):
         logger.info(f"auth failed: {exc.__class__}: {exc} ({auth_token})")
         raise HTTPException(403, "Forbidden")
 
-    # We validate the full url prefix for 2 reasons:
-    # 1) Validating the FQDN prevents possibly use of this token in a different context
-    # 2) All urls start with /workspace/{workspace}/, so it effectively
-    #    constrains a token to a workspace
-    public_url = config.RELEASE_HOST + request.url.path
-    if not public_url.startswith(token.url):
+    validate_url(token, request)
+
+    return token
+
+
+def validate_url(token, request):
+    signed_url = urlparse(token.url)
+
+    # We validate the FQDN in the token is valid for this server. We just
+    # validate the hostname, not the port, as in some backends things end up
+    # going via different ports because $REASONS
```

Comment:
verifying assumptions - this PR whitelists domains, should it also whitelist ports (or even domain/port combos)? or do you think that's overkill? our opinion is that if an attacker could open a port on one of the whitelisted domains, they're already further inside?

Reply:
tl;dr I don't think validating the port adds any security value, and causes code and config complexity.

The primary goal in validating the host in the token is to ensure that the token was intended for this specific backend.  If we just validated the path, then a token could theoretically be replayed against a different backend's release-hatch. This would require the same secret and the same workspace names, which is unlikely, so is already belt-and-braces.

I can't foresee a scenario when we'd have two different release-hatch instances running on the same *domain name* on different ports. The goal is to get rid of non-default ports all together, but we're not there yet. If we ever needed that we'd have different hostnames and use vhosts I think.


Longer I answer that explains more of the motivation, as it's not trivial.

Release-hatch is now running on port 443 in all backends.  In EMIS, the url for the service is therefore `https://emis.backends.opensafely.org` (using local DNS overrides to resolve to the privately accessibly IP when in the right context).
 
Clients need to be configured with a url, a) to connect to and b) to include in the signed token. It's this dual usage purpose that causes problems for us.

The two clients of release-hatch are a) `osrelease`, which runs on the VM, and the outputs-viewer SPA delivered by job-server, which runs in a users browser.

For `osrelease`, if we configure it to use `https://emis.backends.opensafely.org`, and configure release-hatch to use the same, it works, it can connect with and validate the same url.

However, when using output-viewer in their browser, users must use a non-standard port to connect, as they are using SSH forwarding, and SSH forwarding can't use 443 on their local machine. So we configured ssh to do `LocalForward 8001 localhost:443`, so local 8001 is forwarded to 443 on the VM over the SSH connection. So the url they use to connect is `https://emis.backends.opensafely.org:8001`, which works for purpose a) connection but not for purpose b) validation, as release-hatch is configured to accept `https://emis.backends.opensafely.org`, which doesn't match due to the port. And if we add the port in to the release-hatch host config, then osrelease doesn't work as it's not using 8001.

The options to fix are:
 - run release-hatch on 8001 on emis. This would mean special-casing the docker-compose.yaml per backend - it's currently the same everywhere.
 - run release-hatch on 8001 everywhere. I guess this would work but is ugly, and non-443 will likely be problematic when we do want to all access to it in any form. And changing all backends to overcome a weakness just in EMIS networking feels wrong.
 - just validate host not port (this PR).

This allows flexibility about port numbers as needed while still checking that host is correct (which is now less spoofable anyway due to SSL).

Other things, like always allowing for localhost, regardless of port, are useful for testing purposes. e.g. testing inside the docker container, testing `http://localhost:8001` now works.

Longer term, I'm hoping to drop SSH forwarding for tailscale, which means we can just use 443.